pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -124,90 +122,70 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
-func (s *Service) UnregisterEvent() error {
-	return s.disableExecute()
-}
-
-// RegisterMsg on DID Exchange protocol messages. The events are triggered for incoming ConnectionRequest,
-// ConnectionResponse or ConnectionAck message types. The Callback is set to nil in the actionEvent message and service
-// won't be expecting any response from the consumers.
-func (s *Service) RegisterMsg(ch chan<- dispatcher.DIDCommMsg) error {
-	s.statusEventLock.Lock()
-	s.statusEvents = append(s.statusEvents, ch)
-	s.statusEventLock.Unlock()
+	if s.actionEvent != nil {
+		return errors.New("channel is already registered for the action event")
+	}
 
+	s.actionEvent = ch
 	return nil
 }
 
-// UnregisterMsg on DID Exchange protocol messages. Refer RegisterMsg().
-func (s *Service) UnregisterMsg(ch chan<- dispatcher.DIDCommMsg) error {
-	s.statusEventLock.Lock()
-	for i := 0; i < len(s.statusEvents); i++ {
-		if s.statusEvents[i] == ch {
-			s.statusEvents = append(s.statusEvents[:i], s.statusEvents[i+1:]...)
-			i--
-		}
+// UnregisterActionEvent on DID Exchange protocol messages. Refer RegisterActionEvent().
+func (s *Service) UnregisterActionEvent(ch chan<- dispatcher.DIDCommAction) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.actionEvent != ch {
+		return errors.New("invalid channel passed to unregister the action event")
 	}
-	s.statusEventLock.Unlock()
+
+	s.actionEvent = nil
 
 	return nil
 }
 
-// RegisterAutoExecute on DID Exchange protocol messages. When this function is called, the service will auto execute
-// the workflow. This works in conjunction with RegisterEvent(). If this function is called after
-// RegisterEvent(), the service will execute the processing automatically and no action events will be triggered.
-// If both are not set, then service will not handle the messages and throw an error.
-func (s *Service) RegisterAutoExecute() error {
-	s.lock.Lock()
-	s.execute = true
-	s.lock.Unlock()
+// RegisterMsgEvent on DID Exchange protocol messages. The message events are triggered for incoming messages. Service
+// will not expect any callback on these events unlike Action events.
+func (s *Service) RegisterMsgEvent(ch chan<- dispatcher.DIDCommMsg) error {
+	s.msgEventLock.Lock()
+	s.msgEvents = append(s.msgEvents, ch)
+	s.msgEventLock.Unlock()
 
 	return nil
 }
 
-// UnregisterAutoExecute on DID Exchange protocol messages. Refer RegisterAutoExecute().
-func (s *Service) UnregisterAutoExecute() error {
-	return s.disableExecute()
-}
-
-func (s *Service) disableExecute() error {
-	s.lock.Lock()
-	s.actionEvent = nil
-	s.execute = false
-	s.lock.Unlock()
+// UnregisterMsgEvent on DID Exchange protocol messages. Refer RegisterMsgEvent().
+func (s *Service) UnregisterMsgEvent(ch chan<- dispatcher.DIDCommMsg) error {
+	s.msgEventLock.Lock()
+	for i := 0; i < len(s.msgEvents); i++ {
+		if s.msgEvents[i] == ch {
+			s.msgEvents = append(s.msgEvents[:i], s.msgEvents[i+1:]...)
```

Comment:
I think there's a panic here if `ch` is on the last index. Can you include a test for this scenario?

Reply:
Is [this](https://github.com/hyperledger/aries-framework-go/pull/321/files#diff-36ac39bebf32030fd7dff6f12aea8d7eR1003) fine ? Let me know if you see more tests required.

Reply:
Yeah, this feature of Golang is confusing: https://stackoverflow.com/questions/45299185/slice-start-position-greater-than-length-of-the-string